### PR TITLE
[#251] [Connector] [Cloud] [GCP] Add Big Query Support

### LINKED ISSUE #251: [Connector] [Cloud] [GCP] Add Big Query Support
**What feature are your requesting?**

Add Big Query Support in Gimel

**What benefits do you foresee from the feature you are requesting?**

- Cloud Support
- Big Query Support
- 

**Potential solution/ideas?**

**Additional context**

### PR DESCRIPTION
**Pull request by Dee-Pac**
Make sure you have checked all steps below.

### GitHub Issue
Fixes #251 

### Checklist:

- [X] This pull request updates the documentation
- [X] This pull request changes library dependencies
- [X] Title of the PR is of format (example) : [#25][Github] Add Pull Request Template

### What is the purpose of this pull request?

Add Big Query support in Gimel API.

### How was this change validated?

API was validated on Dataproc_1.5 cluster. 

* Start Spark shell 
```bash
spark-shell --jars gs://spark-lib/bigquery/spark-bigquery-latest_2.12.jar,gs://demc-test/lib/gimel-core-2.4.7-SNAPSHOT-uber.jar
```

* Define Hive Catalog as an example

```sql
drop table if exists default.my_dataset1;

create external table default.my_dataset1
(
cols String
)
location "gs://demc-test/pp-devcos-dataproc-gcs.BQ_benchmark.date_dim1"
TBLPROPERTIES(
'gimel.storage.type'='bigquery',
'temporaryGcsBucket'='demc-test',
'gimel.bigquery.table'='pp-devcos-dataproc-gcs.BQ_benchmark.date_dim1'
)
```

* Read API

```scala
val dataset = com.paypal.gimel.DataSet(spark)
spark.sql("set gimel.catalog.provider=HIVE")

val df1 = dataset.read("default.my_dataset1")

df1.show()

+---------+----------------+----------+-----------+----------+-------------+------+-----+-----+-----+-----+---------+----------------+-------------+----------+--------------+---------+---------+-------------------+-----------+----------+-------------+-------------+-------------+--------------+---------------+-----------------+--------------+
|d_date_sk|       d_date_id|    d_date|d_month_seq|d_week_seq|d_quarter_seq|d_year|d_dow|d_moy|d_dom|d_qoy|d_fy_year|d_fy_quarter_seq|d_fy_week_seq|d_day_name|d_quarter_name|d_holiday|d_weekend|d_following_holiday|d_first_dom|d_last_dom|d_same_day_ly|d_same_day_lq|d_current_day|d_current_week|d_current_month|d_current_quarter|d_current_year|
+---------+----------------+----------+-----------+----------+-------------+------+-----+-----+-----+-----+---------+----------------+-------------+----------+--------------+---------+---------+-------------------+-----------+----------+-------------+-------------+-------------+--------------+---------------+-----------------+--------------+
|  2452641|AAAAAAAABKMGFCAA|2003-01-01|       1236|      5375|          413|  2003|    3|    1|    1|    1|     2003|             413|         5375| Wednesday|        2003Q1|        Y|        N|                  N|    2452641|   2452640|      2452276|      2452549|            N|             N|              Y|                Y|             Y|
+---------+----------------+----------+-----------+----------+-------------+------+-----+-----+-----+-----+---------+----------------+-------------+----------+--------------+---------+---------+-------------------+-----------+----------+-------------+-------------+-------------+--------------+---------------+-----------------+--------------+

```

* Write API

```scala

dataset.write("default.my_dataset1",df1)

```

* Validate Read (after write)
```scala

dataset.read("default.my_dataset1").show()

+---------+----------------+----------+-----------+----------+-------------+------+-----+-----+-----+-----+---------+----------------+-------------+----------+--------------+---------+---------+-------------------+-----------+----------+-------------+-------------+-------------+--------------+---------------+-----------------+--------------+
|d_date_sk|       d_date_id|    d_date|d_month_seq|d_week_seq|d_quarter_seq|d_year|d_dow|d_moy|d_dom|d_qoy|d_fy_year|d_fy_quarter_seq|d_fy_week_seq|d_day_name|d_quarter_name|d_holiday|d_weekend|d_following_holiday|d_first_dom|d_last_dom|d_same_day_ly|d_same_day_lq|d_current_day|d_current_week|d_current_month|d_current_quarter|d_current_year|
+---------+----------------+----------+-----------+----------+-------------+------+-----+-----+-----+-----+---------+----------------+-------------+----------+--------------+---------+---------+-------------------+-----------+----------+-------------+-------------+-------------+--------------+---------------+-----------------+--------------+
|  2452641|AAAAAAAABKMGFCAA|2003-01-01|       1236|      5375|          413|  2003|    3|    1|    1|    1|     2003|             413|         5375| Wednesday|        2003Q1|        Y|        N|                  N|    2452641|   2452640|      2452276|      2452549|            N|             N|              Y|                Y|             Y|
|  2452641|AAAAAAAABKMGFCAA|2003-01-01|       1236|      5375|          413|  2003|    3|    1|    1|    1|     2003|             413|         5375| Wednesday|        2003Q1|        Y|        N|                  N|    2452641|   2452640|      2452276|      2452549|            N|             N|              Y|                Y|             Y|
+---------+----------------+----------+-----------+----------+-------------+------+-----+-----+-----+-----+---------+----------------+-------------+----------+--------------+---------+---------+-------------------+-----------+----------+-------------+-------------+-------------+--------------+---------------+-----------------+--------------+

```

* Validate options are working

```scala

val options = Map("saveMode" -> "ErrorIfExists")

// Write must Error out ( as expected )
dataset.write("default.my_dataset1",res7,options)

```

Output
`java.lang.IllegalArgumentException: SaveMode is set to ErrorIfExists and Table pp-devcos-dataproc-gcs.BQ_benchmark.date_dim1 already exists. Did you want to add data to the table by setting the SaveMode to Append? Example: df.write.format.options.mode(SaveMode.Append).save()`

### Commit Guidelines
- [X] My commits all reference GH issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Review by bthilakraj — COMMENTED** [review]
1.How is cloud storage bucket props being set?i dont see that in the PR or am i missing anything.please clarify.
2.Can you also pls include the test plan for this diff and attach screenshots if any to validate the results?

**Comment by Dee-Pac** [comment]
@bthilakraj  Thanks for the comments. Addressed those.
I'll update this PR once further testing is complete & ready for final review.

**Comment by Dee-Pac** [comment]
@bthilakraj - 

- I've addressed the review comments.
- Also added snippets validated on the dataproc cluster

File: gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/utilities/BigQueryUtilities.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * limitations under the License.
+ */
+
+package com.paypal.gimel.bigquery.utilities
+
+import com.paypal.gimel.common.conf.GimelConstants
+
+object BigQueryUtilities {
+
+
```
[the remark this answers is not in the document]
Comment (Dee-Pac) [reply gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/utilities/BigQueryUtilities.scala:64]: It was meant for future in the initial commit, now it has a few reusable utility methods.

File: gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/utilities/BigQueryUtilities.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  val logger = Logger()
+
+  /**
+   * Check if Big Query Table name is supplied. If not supplied : fail.
+   * @param bigQueryTable The Big Query Table Name in the form of "gimel.bigquery.table"
+   */
+  def failIfTableNotSpecified(bigQueryTable: Option[String]): Unit = {
+    bigQueryTable match {
+      case Some(tbl) => logger.info(s"Found Table ${bigQueryTable.get}")
+      case _ => throw new Exception(
```
Comment (sathishsri88) [inline gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/utilities/BigQueryUtilities.scala:37]: Can we throw specific exception like IllegalArgumentException

File: gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/utilities/BigQueryUtilities.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    saveMode match {
+      case BigQueryConstants.saveModeAppend =>
+        logger.info(s"Appending to Big Query table [${bigQueryTable}]")
+      case BigQueryConstants.saveModeErrorIfExists =>
+        logger.info(s"Asked to Err if exists Big Query table [${bigQueryTable}]")
+      case BigQueryConstants.saveModeOverwrite =>
+        logger.info(s"Overwriting Big Query table [${bigQueryTable}]")
+      case BigQueryConstants.saveModeIgnore =>
+        logger.info(s"Asked to Ignore if exists Big Query table [${bigQueryTable}]")
+      case _ => throw new Exception(
```
Comment (sathishsri88) [inline gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/utilities/BigQueryUtilities.scala:57]: Can we throw specific exception like IllegalArgumentException

File: gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/DataSet.scala [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+      utils.parseSaveMode(saveMode, bigQueryTable.get)
+      val outDf: DataFrameWriter[Row] = dataFrame.write.format(BigQueryConstants.bigQuery)
+      logger.info("Applying all the supplied options from catalog + runtime to the bigquery connector...")
+      logger.info(options.mkString("\n"))
+      val dfWriter: DataFrameWriter[Row] = options.foldLeft(outDf)((initDfReader, each) => initDfReader.option(each._1, each._2))
+      dfWriter.mode(saveMode).save(bigQueryTable.get)
+      dataFrame
+    } catch {
+      case e: Throwable =>
+        logger.info(s"${BigQueryConstants.bigQueryDocUrl}")
```
Comment (sathishsri88) [inline gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/DataSet.scala:103]: can we change it to `logger.error` with the exception message

File: gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/DataSet.scala [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+      val bigQueryTable: Option[String] = options.get(BigQueryConfigs.bigQueryTable)
+      utils.failIfTableNotSpecified(bigQueryTable)
+      val initDfReader: DataFrameReader = sparkSession.read.format(BigQueryConstants.bigQuery)
+      logger.info("Applying all the supplied options from catalog + runtime to the bigquery connector...")
+      logger.info(options.mkString("\n"))
+      val dfReader: DataFrameReader = options.foldLeft(initDfReader)((initDfReader, each) => initDfReader.option(each._1, each._2))
+      dfReader.load(bigQueryTable.get)
+    } catch {
+      case e: Throwable =>
+        logger.info(s"${BigQueryConstants.bigQueryDocUrl}")
```
Comment (sathishsri88) [inline gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/DataSet.scala:68]: can we change it to logger.error with the exception message

File: gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/conf/BigQueryConstants.scala [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.paypal.gimel.bigquery.conf
+
+object BigQueryConstants {
+  val bigQuery: String = "bigquery"
```
Comment (sathishsri88) [inline gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/conf/BigQueryConstants.scala:23]: As a general practice, Constants we have uppercase with words separated by underscores.

Comment (Dee-Pac) [reply gimel-dataapi/gimel-connectors/gimel-bigquery/src/main/scala/com/paypal/gimel/bigquery/conf/BigQueryConstants.scala:23]: @sathishsri88  We have this pattern across all the connectors and codebase. I would revisit this standard as a larger refactoring PR.

**Comment by Dee-Pac** [comment]
@sathishsri88  - I've addressed all your comments. Could you take a look?

**Review by bthilakraj — COMMENTED** [review]
@Dee-Pac PR LGTM, except one question -> why are we deleting docs related kafka, hbase, druid as part of this PR(This is beyond the scope of this PR Title)?
I want to make sure so that i dont want merge it and miss any useful documentation.
After your confirmation on this i can approve and merge this PR as code looks good to me.

**Comment by Dee-Pac** [comment]
> @Dee-Pac PR LGTM, except one question -> why are we deleting docs related kafka, hbase, druid as part of this PR?(This is beyond the scope of this PR Title)
> 
> Screen Shot 2020-11-22 at 12 00 49 AM
> 
> I want to make sure, so that i dont want to merge and miss any useful documentation.
> After your confirmation on this i can approve and merge this PR as code looks good to me.

@bthilakraj Thanks for asking this question. I missed to provide clarity while raising the PR.

This PR is merging into the dataproc branch. The reason is : the connectors I've excluded as part of this PR are not compatible with the spark version or the scala versions.
We need to tackle each connector separately.

The idea is to gradually cherry-pick each connector, upgrade the version , push to dataproc branch.

Eventually merge the dataproc branch to main branch.

hope this clarifies.

**Comment by bthilakraj** [comment]
@Dee-Pac Sounds good.Thanks for corfirming

**Pull request merged**